Actually check the CollectCom transition consumes consistent PTs/STs

KtorZ opened the pull request:
Fixes #241

KtorZ committed:
Mutate ST's head id in collect-com => adversarial mutation

  The idea is to cause a mutation to force the inspection of
  participation tokens in the collect-com. Currently, tokens are only
  counted (and even... we only truly count commits). Yet, we do
  want in the end to look for token which have the same policy id as the
  thread token, and interpret them as PTs.

  Thus, by mutating the policy id of the thread token, we should in
  principle invalidate all PTs based on a different policy id. Which,
  for now, does not happen, hence the failing mutation test.

KtorZ committed:
Extract thread token from context in collect-com transition.

abailly committed:
Extract participation tokens from commits #241

KtorZ committed:
Make healthy collect-com healthy again by adding PTs to commits.

KtorZ committed:
Adjust API for 'ChangeInput' to allow passing a redeemer.

abailly committed:
Fix redeemer update to keep right data when Head Id changes #241

abailly committed:
Reduce number of commits (budget overspent) #241

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -440,17 +441,23 @@ alterRedeemers fn = \case
      in TxBodyScriptData dats (Ledger.Redeemers newRedeemers)
 
 -- | Remove redeemer for given `TxIn` from the transaction's redeemers map.
```
Should: update this outdated doc.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Commit.hs:
```diff
@@ -80,11 +80,12 @@ genCommitMutation (tx, _utxo) =
         mutatedValue <- genValue `suchThat` (/= committedOutputValue)
         let mutatedOutput = modifyTxOutValue (const mutatedValue) committedTxOut
-        pure $ ChangeInput committedTxIn mutatedOutput
+
```
Could: remove redundant new line

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -126,21 +129,37 @@ checkCollectCom commitAddress (_, parties) context@ScriptContext{scriptContextTx
+      & filter (/= adaSymbol)
+      & \case
+        [headPolicyId] -> headPolicyId
+        _ -> traceError "malformed thread token, expected single asset"
```
Nice code. We should use consistent names though.

Should: As it seems not to be the thread token we retrieve, but this head's currency symbol, please rename `threadToken`.

Could: More minor, the term `headPolicyId` is not consistent with what plutus is calling it (ref `symbols` and `adaSymbol`). So maybe we should also be consistent in that regards?

KtorZ replied:
> As it seems not to be the thread token we retrieve, but this head's currency symbol

Yup, I told myself the same thing yesterday when I self reviewed the code. It's not a thread token indeed.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -145,6 +168,13 @@ genCollectComMutation (tx, utxo) =
             , ChangeOutput 0 $ mutatedPartiesHeadTxOut moreParties
             ]
+    , SomeMutation MutateHeadId <$> do
+        illedHeadResolvedInput <-
```
:joy: is that even a word? Maybe `sickHeadResolvedInput` is also an option?

No need to change, though.

KtorZ replied:
I thought that as a German-speaker, you would appreciate our creativity on words :nerd_face:

ch1bo reviewed (approved):
Only some comments on naming things and one outdated haddock.

KtorZ committed:
Count commits with PTs only

  Make validator more efficient by avoiding constructing and traversing
  a list of PTs. And, this also reflects a bit more the logic itself.

KtorZ committed:
Add participation tokens to collect-com tx template

  So that, collect-com can validate properly.. but, we are running out of budget.

KtorZ committed:
Reduce the number of participant down to 3 in collect-com.

  We need to optimize the collect-com! ... and probably other validators too.

KtorZ committed:
Factor out fanout check as a separate function.

  Fortunately, this is a cost-free change! :tada:.

KtorZ committed:
Slightly increase maximum memory budget for end-to-end

  This is a temporary measure because, if we indeed want to run this on
  THE testnet, we need the execution budget to be smaller. So optimizing
  collect-com is a must (or reviewing the approach to do smaller steps
  via sequential collect of commits).

  → #254

ch1bo committed:
Fix genAbortableOutputs to not generate overlapping TxIns

Discovered in a flaky test which saw didn't see enough initials/commits
because their TxIns were conflicting

ch1bo committed:
Fix Arbitrary TxIn and add regression properties

Arbitrary TxIn, TxId, VerificationKey and Hashes thereof should not
conflict as this would lead to surprising test failures
